Dockerfile

```diff
@@ -0,0 +1,58 @@
+ARG ALPINE_VERSION="3.12"
+ARG PYTHON_VERSION="3.9"
+ARG BASE_IMAGE="python:${PYTHON_VERSION}-alpine${ALPINE_VERSION}"
+
+# Stage 1 - bundle base image + runtime
+FROM ${BASE_IMAGE} AS python-alpine 
+RUN apk add \
+    --no-cache \
+    libstdc++ 
+
+# Stage 2 - build function and dependencies
+FROM python-alpine AS build-image
+
+RUN apk add \
+    --no-cache \ 
+    build-base \
+    libtool \
+    autoconf \
+    automake \
+    libexecinfo-dev \
+    make \
+    cargo \
+    rust \
+    cmake \
+    libcurl \
+    libffi-dev \
+    openssl-dev \
+    groff 
+
+
+RUN python -m venv /opt/venv
+ENV PATH="/opt/venv/bin:$PATH" 
+RUN python3 -m pip install awslambdaric==1.0.0 boto3==1.17
+
+COPY . .
+RUN python3 -m pip install -r requirements.txt
+
+# Stage 3 - final runtime image
+FROM python-alpine
+WORKDIR /home
+
+RUN apk add \
+    --no-cache \
+    openssl \
+    openjdk8-jre
+
+COPY --from=build-image /opt/venv /opt/venv
+ENV PATH="/opt/venv/bin:$PATH"
+
+COPY --from=build-image heracles ./heracles
+
+#(Optional) Add Lambda Runtime Interface Emulator and use a script in the ENTRYPOINT for simpler local runs
```

Comment:
Why not just use the official `public.ecr.aws/lambda/python:3.9` image? It includes the emulator by default and should have everything we need to run this code. In addition, using that would probably cut the Dockerfile down to 10-15 lines.

Reply:
You're right, the complexity of the Dockerfile would be way less:

I tested with this and it also works:

```
FROM public.ecr.aws/lambda/python:3.8

COPY requirements.txt .
RUN python3 -m pip install -r requirements.txt && rm requirements.txt

COPY heracles ./heracles

CMD [ "heracles.lambda.handler" ]
```

But the image is almost triple the size at about 600mb. I like smaller images, but I agree it adds a lot of complexity to the Dockerfile. So, your call :).

By the way, the 3.9 tag is not yet released for the base image.

Reply:
@dacort I changed it now to use the official base image for easier maintainability.